src/sw/to-cache.ts

```diff
@@ -0,0 +1,213 @@
+import * as wp2EncMtSimd from 'entry-data:codecs/wp2/enc/wp2_enc_mt_simd';
+import wp2EncMtSimdWasm from 'url:codecs/wp2/enc/wp2_enc_mt_simd.wasm';
+import * as wp2EncMtWorker from 'entry-data:codecs/wp2/enc/wp2_enc_mt.worker.js';
+import * as wp2EncMt from 'entry-data:codecs/wp2/enc/wp2_enc_mt';
+import wp2EncMtWasm from 'url:codecs/wp2/enc/wp2_enc_mt.wasm';
+import * as wp2Enc from 'entry-data:codecs/wp2/enc/wp2_enc';
+import wp2EncWasm from 'url:codecs/wp2/enc/wp2_enc.wasm';
+
+export const theRest = (async () => {
+  const [
+    supportsThreads,
+    supportsSimd,
+    supportsWebP,
+    supportsAvif,
+  ] = await Promise.all([
+    threads(),
+    simd(),
+    ...[webpDataUrl, avifDataUrl].map(async (dataUrl) => {
+      if (!self.createImageBitmap) return false;
+      const response = await fetch(dataUrl);
+      const blob = await response.blob();
+      return createImageBitmap(blob).then(
+        () => true,
+        () => false,
+      );
+    }),
+  ]);
+
+  const items = [
+    featuresWorker.main,
+    ...featuresWorker.deps,
+    rotateWasm,
+    quantWasm,
+    resizeWasm,
+    hqxWasm,
+    mozjpegWasm,
+    jxlDecWasm,
+    wp2DecWasm,
+  ];
+
+  if (!supportsAvif) items.push(avifDec.main, ...avifDec.deps, avifDecWasm);
+  if (!supportsWebP) items.push(webpDec.main, ...webpDec.deps, webpDecWasm);
+
+  // AVIF
+  if (supportsThreads) {
+    items.push(
+      avifEncMtWorker.main,
+      ...avifEncMtWorker.deps,
+      avifEncMt.main,
+      ...avifEncMt.deps,
+      avifEncMtWasm,
+    );
+  } else {
+    items.push(avifEnc.main, ...avifEnc.deps, avifEncWasm);
+  }
+
+  // JXL
+  if (supportsThreads && supportsSimd) {
+    items.push(
+      jxlEncMtSimdWorker.main,
+      ...jxlEncMtSimdWorker.deps,
+      jxlEncMtSimd.main,
+      ...jxlEncMtSimd.deps,
+      jxlEncMtSimdWasm,
+    );
+  } else if (supportsThreads) {
+    items.push(
+      jxlEncMtWorker.main,
+      ...jxlEncMtWorker.deps,
+      jxlEncMt.main,
+      ...jxlEncMt.deps,
+      jxlEncMtWasm,
+    );
+  } else {
+    items.push(jxlEnc.main, ...jxlEnc.deps, jxlEncWasm);
+  }
+
+  // OXI
+  if (supportsThreads) {
+    items.push(oxiMtWorker.main, ...oxiMtWorker.deps, oxiMtWasm);
+  } else {
+    items.push(oxiWasm);
+  }
+
+  // WebP
+  if (supportsSimd) {
+    items.push(webpEncSimd.main, ...webpEncSimd.deps, webpEncSimdWasm);
+  } else {
+    items.push(webpEnc.main, ...webpEnc.deps, webpEncWasm);
+  }
+
+  // WP2
+  if (supportsThreads && supportsSimd) {
+    items.push(
+      wp2EncMtSimdWorker.main,
+      ...wp2EncMtSimdWorker.deps,
+      wp2EncMtSimd.main,
+      ...wp2EncMtSimd.deps,
+      wp2EncMtSimdWasm,
+    );
+  } else if (supportsThreads) {
+    items.push(
+      wp2EncMtWorker.main,
+      ...wp2EncMtWorker.deps,
+      wp2EncMt.main,
+      ...wp2EncMt.deps,
+      wp2EncMtWasm,
+    );
+  } else {
+    items.push(wp2Enc.main, ...wp2Enc.deps, wp2EncWasm);
+  }
+
+  return [...new Set(items)];
+})();
```

Comment:
This is a much more manual process than we had with webpack. WIth webpack we were making guesses based on file names. This is more of a 'proper' way, although it requires maintenance when we add new files.

The string-guessing way wouldn't scale once we had complex feature detection stuff like we have here.

Reply:
I like this. It’s a bit of code but it’s very clear.

(At some point I’d like Emscripten to support developers here better)